### PR TITLE
USIO: Support up to 4 players for Tekken Tag Tournament 2 Pair Play mode

### PR DESCRIPTION
**Pull request by brian218**
Support up to 4 players for Tekken Tag Tournament 2 Pair Play mode.

SCEEXE000_screenshot_2023_09_06
SCEEXE000_screenshot_2023_09_06
SCEEXE000_screenshot_2023_09_06
SCEEXE000_screenshot_2023_09_06
SCEEXE000_screenshot_2023_09_06
Image

### CODE REVIEW

**Comment by brian218** [comment]
Reference: original emulation introduced in #14253 only supports up to 2 players.

File: rpcs3/Emu/Io/usio.h [hunk 1]
```diff
@@ -21,11 +21,11 @@ class usb_device_usio : public usb_device_emulated
...
 
 private:
-	bool test_on = false;
-	bool test_key_pressed = false;
-	bool coin_key_pressed = false;
+	bool test_on[2]{};
+	bool test_key_pressed[2]{};
+	bool coin_key_pressed[2]{};
+	le_t<u16> coin_counter[2]{};
 	bool is_used = false;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.h:28]: I'd rather have a std array of two structs containing these as members

Comment (Megamouse) [reply rpcs3/Emu/Io/usio.h:28]: (The four values that were duplicated)

Comment (brian218) [reply rpcs3/Emu/Io/usio.h:24]: Done.

File: rpcs3/Emu/Io/usio.cpp [hunk 2]
```diff
@@ -212,15 +212,15 @@ void usb_device_usio::translate_input_taiko()
 				{
 				case usio_btn::test:
 					if (player != 0) break;
-					if (pressed && !test_key_pressed) // Solve the need to hold the Test key
-						test_on = !test_on;
-					test_key_pressed = pressed;
+					if (pressed && !per_io_status[0].test_key_pressed) // Solve the need to hold the Test key
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.cpp:215]: Just do
`io_status& status = per_io_status[0];`
And use that below. Then you don't have to access the correct index in every following usage.

Same goes for the Tekken class.

File: rpcs3/Emu/Io/usio.h [hunk 3]
```diff
@@ -21,11 +21,17 @@ class usb_device_usio : public usb_device_emulated
...
+
+	struct io_status
+	{
+		bool test_on = false;
+		bool test_key_pressed = false;
+		bool coin_key_pressed = false;
+		le_t<u16> coin_counter = 0;
+	};
+
+	std::array<io_status, 2> per_io_status;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.h:36]: m_per_io_status

File: rpcs3/Emu/Io/usio.cpp (hunk 2 above)
Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Done. Thanks!

File: rpcs3/Emu/Io/usio.cpp [hunk 4]
```diff
@@ -292,115 +295,119 @@ void usb_device_usio::translate_input_tekken()
...
-					coin_key_pressed = pressed;
+					if (pressed && !status.coin_key_pressed) // Ensure only one coin is inserted each time the Coin button is pressed
+						status.coin_counter++;
+					status.coin_key_pressed = pressed;
 					break;
 				case usio_btn::service:
-					if (player == 0 && pressed)
-						digital_input |= 0x4000;
+					if (player % 2 == 0 && pressed)
+						digital_input[player / 2] |= 0x4000;
```
Comment (Megamouse) [inline rpcs3/Emu/Io/usio.cpp:322]: Same with this one...

Comment (brian218) [reply rpcs3/Emu/Io/usio.cpp]: Done.

**Pull request merged**